### DISCUSSION TITLE
Add `sourceUrlTemplate` to customManagers

### DISCUSSION THREAD

**Discussion by jonasbadstuebner** [renovatebot/renovate#27930] [Suggest an Idea]
### Tell us more.

I have a use case where I have the following self hosted GitLab setup:
```
GitLab: git.example.com
Project: team/developers
Release/Tag: my-release-1.0.0
sourceDirectory: packages/my-release
```

So the Tag/Release (I let both be created) is at `https://git.example.com/team/developers` and called `my-release-1.0.0` and source code and CHANGELOG.md are at `packages/my-release` in `https://git.example.com/team/developers`.

Renovate does not give me an option to set the sourceUrl for gitlab-tags to something else than the package name. But the package name has to be `team/developers`.

So I had the idea to add `sourceUrlTemplate` to customManagers. Theoretically it should be possible, right?
https://github.com/renovatebot/renovate/blob/8fc50ca86f3de5b29a99d1ac0263d3030a46e085/lib/modules/manager/types.ts#L126
->
https://github.com/renovatebot/renovate/blob/8fc50ca86f3de5b29a99d1ac0263d3030a46e085/lib/modules/manager/custom/regex/utils.ts#L63

So adding it here https://github.com/renovatebot/renovate/blob/8fc50ca86f3de5b29a99d1ac0263d3030a46e085/lib/modules/manager/custom/regex/utils.ts#L13 should be enough, right?

**Comment by jonasbadstuebner** [comment]
Maybe there is a different approach, but I don't know how I can give Renovate the correct path to the CHANGELOG.md in my monorepo.

**Comment by rarkins** [comment]
We don't want to overload custom managers. Why can't it be automatically determined by the GitLab-tags datasource?

**Reply by jonasbadstuebner** [reply]
If I read the source code of the datasource correctly, it does not allow me to set the sourceDirectory manually. I can change the lookup path by changing the packageName, which implies that CHANGELOG.md is always at the root of my GitLab project.
https://github.com/renovatebot/renovate/blob/8fc50ca86f3de5b29a99d1ac0263d3030a46e085/lib/modules/datasource/gitlab-tags/util.ts#L10-L16

**Reply by jonasbadstuebner** [reply]
I thought we could add it to custom managers, so it is available for all datasources.

**Reply by rarkins** [reply]
sourceUrl is a datasource concept and wrong to put it in a manager.

**Reply by jonasbadstuebner** [reply]
The helm datasource for example has a `findSourceUrl` method. For GitLab (and probably a lot of others) this does not exist, afaik.

What would be your preferred way of specifying a sourceUrl for GitLab-tags?

**Reply by rarkins** [reply]
I think renovate can try to hunt in packages/depName and packages/packageName folders for example

**Reply by jonasbadstuebner** [reply]
`packages` is just an example in my case. So it would have to be configurable.
Also, `packageName` for Gitlab-Tags is the repository in GitLab.
If I have `https://gitlab.com/renovatebot/renovate`, `packageName` would have to be `renovatebot/renovate`. If there was a subfolder, where would I specify this?
I cannot change `packageName`, because Renovate uses `https://gitlab.com/api/v4/projects/$urlEncodedPackageName/repository/tags?per_page=100` to look up the tags.

**Reply by jonasbadstuebner** [reply]
The way I understand it, managers are the way to tell datasources what to do/where to look/what to look for, right?

If I cannot "extend" the existing datasource and I cannot reimplement the GitLab-Tags datasource with plain Renovate-JSON-Config, there should be a way to specify or change the resulting `sourceUrl` through a custom manager.

**Reply by rarkins** [reply]
I don't think we want to add this complexity for what seems like an edge case. SourceUrl isn't the right way to do it anyway, this concept is sourceDirectory.

**Reply by jonasbadstuebner** [reply]
What is the way to tell GitLab the sourceDirectory?

**Reply by rarkins** [reply]
Manager extracts list of dependencies to look up from datasource. Each datasource can optionally tell us where to find the source repository/source directory from. You can find examples of this in datasources searching for sourceUrl and sourceDirectory.

If sourceDirectory is not specified but sourceUrl is then Renovate still does some effort during changelog fetching to look inside subdirectories, but there needs to be some structure/convention. Renovate won't be able to guess if (a) no sourceDirectory, and (b) random, non-conventional directory naming.

**Reply by jonasbadstuebner** [reply]
I agree that this is a special use case.
Not worth implementing it upstream.

I still did not get the changelog to work. And maybe you can give me a hint, why?
I have a custom datasource that's giving me an output similar to the one below (the commit SHAs are just for show)
```
{
  "releases": [
    {
      "version": "0.10.0",
      "sourceUrl": "https://git.example.com/team/developers/-/tree/1a2b3c4d",
      "sourceDirectory": "packages/packageA",
      "changelogUrl": "https://git.example.com/team/developers/-/blob/1a2b3c4d/packages/packageA/CHANGELOG.md",
      "gitRef": "packageA-0.10.0",
      "releaseTimestamp": "2024-03-11T09:40:00.000+00:00"
    },
    [ other releases ]
  ]
}
```
I enabled debug log level, but I can not see, that Renovate even checks any of this.
In the MergeRequest it shows, that there was a problem with looking up dependencies.

Is any of my keys wrong? Or does the custom datasource not support changelogs? (According to the documentation, it should work like this.)

**Reply by jonasbadstuebner** [reply]
Making the changelog url point to the raw markdown file (instead of using "blob") also did not work.

**Reply by rarkins** [reply]
Try trimming the sourceUrl to stop at the project

**Reply by jonasbadstuebner** [reply]
Unfortunately, this did not solve anything. When I set the log level to trace, I have renovate crashing, so I cannot investigate very well, where it tries to grab the changelog from.
From the documentation I would read it that I can set the `changelogUrl`:
>      "changelogUrl": "https://github.com/demo-org/demo/blob/main/CHANGELOG.md#v0710",
So I would guess that I could set the `sourceUrl` and `sourceDirectory` to whatever and it would still work. But it doesn't.

**Reply by rarkins** [reply]
We're looking into whether we can make sourceUrl configurable by users without it interfering with caching/other users

**Reply by jonasbadstuebner** [reply]
Is the sourceUrl really the problem?
I am wondering if I‘m doing something wrong. Shouldn’t the changelogUrl pointing to the Changelog.md be enough for renovate to fetch all informiation needed?

**Reply by rarkins** [reply]
I'm not sure without being able to run it and take a look myself, but a reproduction will be hard in this case due to it being a private endpoint
